src/main/java/com/ontotext/trree/plugin/mongodb/MongoResultIterator.java

```diff
@@ -512,60 +513,66 @@ public StatementIterator getModelIterator(final long subject, final long predica
-					if (!initialized && !initializedByEntityIterator) {
-						if (!isQuerySet()) {
-							return false;
-						}
-						initializedByEntityIterator = true;
-						if (initialize()) {
-							advance();
-						} else {
-							// no solutions were found or could not connect
-							return false;
-						}
-					} else {
-						if (hasSolution()) {
-							advance();
+				while (true) {
+					if (currentRDF == null) {
+						if (!initialized && !initializedByEntityIterator) {
+							if (!isQuerySet()) {
+								return false;
+							}
+							initializedByEntityIterator = true;
+							if (initialize()) {
+								advance();
+							} else {
+								// no solutions were found or could not connect
+								return false;
+							}
 						} else {
-							// no more solutions
-							return false;
+							if (hasSolution()) {
+								advance();
+							} else {
+								// no more solutions
+								return false;
+							}
 						}
 					}
-				}
-				if (local == null) {
-					// see the comment above
-					Resource localSub = s;
-					if (localSub == null && batched) {
-						localSub = (Resource) entities.get(MongoResultIterator.this.object);
-						if (localSub != null && localSub.equals(o)) {
-							localSub = null;
+					if (local == null) {
+						// see the comment above
+						Resource localSub = s;
+						if (localSub == null && batched) {
+							localSub = (Resource) entities.get(MongoResultIterator.this.object);
+							if (localSub != null && localSub.equals(o)) {
+								localSub = null;
+							}
 						}
+						local = currentRDF.filter(localSub, p, o).iterator();
 					}
-					local = currentRDF.filter(localSub, p, o).iterator();
-				}
-				boolean has = local.hasNext();
-//				if (!has && hasSolution()) {
-//					advance();
-//					local = currentRDF.filter(s, p, o).iterator();
-//					has = local.hasNext();
-//				}
-				if (has) {
-					Statement st = local.next();
-					this.subject = entities.resolve(st.getSubject());
-					if (this.subject == 0) {
-						this.subject = entities.put(st.getSubject(), Scope.REQUEST);
+					if (local.hasNext()) {
+						Statement st = local.next();
+						this.subject = entities.resolve(st.getSubject());
+						if (this.subject == 0) {
+							this.subject = entities.put(st.getSubject(), Scope.REQUEST);
+						}
+						this.predicate = entities.resolve(st.getPredicate());
+						if (this.predicate == 0) {
+							this.predicate = entities.put(st.getPredicate(), Scope.REQUEST);
+						}
+						this.object = entities.resolve(st.getObject());
+						if (this.object == 0) {
+							this.object = entities.put(st.getObject(), Scope.REQUEST);
+						}
+						return true;
 					}
-					this.predicate = entities.resolve(st.getPredicate());
-					if (this.predicate == 0) {
-						this.predicate = entities.put(st.getPredicate(), Scope.REQUEST);
+
+					// currentRDF exhausted, try next document/batch
+					local = null;
+					if (entityIteratorCreated) {
+						return false;
```

Comment:
what is the effect of this change here?

Does it fix the problem with Graph redirection or it's something else?